### ISSUE TITLE
libpng 1.6.36 png_set_filter() wrong mask with PNG_FILTER_VALUE_*

### ISSUE THREAD

**Issue by a-kulakov** [glennrp/libpng#262]
libpng 1.6.36 png_set_filter() resulting filter mask for PNG_FILTER_VALUE_* differ from mask, produced by libpng 1.2 and libpng 1.5

Test on linux can be build with cmake and run with ```./parent_dir_build.sh``` in
https://github.com/a-kulakov/libpng_set_filter_check
Current result on Ubuntu 16.04 with
- PNG_FILTER_VALUE_NONE
- PNG_FILTER_VALUE_SUB
- PNG_NO_FILTERS (==PNG_FILTER_VALUE_NONE)
```
-rwxrwxr-x 1 113216 Dec  4 02:39 _build/libpng12/set_filter_check
-rwxrwxr-x 1 133712 Dec  4 02:39 _build/libpng15/set_filter_check
-rwxrwxr-x 1 318688 Dec  4 02:39 _build/libpng16/set_filter_check
libpng version 1.2.59
 in v1.2 out (hex values, v1.2 - libpng12 output value)
  0  08  08 
  1  10  10 
  0  08  08 
libpng version 1.5.30
 in v1.2 out (hex values, v1.2 - libpng12 output value)
  0  08  08 
  1  10  10 
  0  08  08 
libpng version 1.6.36
 in v1.2 out (hex values, v1.2 - libpng12 output value)
  0  08  00 !=
  1  10  01 !=
  0  08  00 !=
```
Unexpected mask changes can increase resulting png size, affect performance, etc.

**Comment by a-kulakov** [comment]
```BRANCHES='libpng-1.6.17-signed libpng-1.6.18-signed libpng17' ../libpng_set_filter_check/parent_dir_build.sh
-rwxrwxr-x 1 317640 дек  4 18:36 _build/libpng-1.6.17-signed/set_filter_check
-rwxrwxr-x 1 305336 дек  4 18:35 _build/libpng-1.6.18-signed/set_filter_check
-rwxrwxr-x 1 405696 дек  4 15:57 _build/libpng17/set_filter_check
libpng version 1.6.17
 in v1.2 out (hex values, v1.2 - libpng12 output value)
  0  08  08 
  1  10  10 
  0  08  08 
libpng version 1.6.18
 in v1.2 out (hex values, v1.2 - libpng12 output value)
  0  08  00 !=
  1  10  01 !=
  0  08  00 !=
libpng version 1.7.0beta90
 in v1.2 out (hex values, v1.2 - libpng12 output value)
  0  08  08 
  1  10  10 
  0  08  08 
```

**Comment by ctruta** [comment]
The regression was most likely introduced at the time when the weighted filter selection code was removed. This happened between libpng-1.6.17 and libpng-1.6.18, as confirmed in the bug report.

This is a great find, Anatoli. Thank you very much for the exceptional testing and bisecting work!
Спасибо.

**Comment by a-kulakov** [comment]
Cosmin, thanks.
We can discuss commit in https://github.com/glennrp/libpng/pull/263

Somewhat later I'll try to modify libpng_set_filter_check.
It looks like multiple set_filter() have side effects and to check code it's better create png_struct for each filter check.

**Comment by a-kulakov** [comment]
@ctruta png_set_filter() can set PNG_FILTER_NONE on undocumented filters parameter values.
I believe this shall lead to more reliable and predictable code.
Link to branch with working code:
https://github.com/a-kulakov/libpng/tree/libpng16-set_filter-undocumented-sets-none-filter

If you think that this idea have sense, I'll remove first commit, rewrite history and can make another pull request.
If required or desirable I'll try to implement same behaviour in libpng17 branch.

**Comment by a-kulakov** [comment]
Ok, another branch is ready for pull request to replace https://github.com/glennrp/libpng/pull/263 pull request.
https://github.com/a-kulakov/libpng/tree/libpng16-set_filter-undocumented-sets-none-filter
I'm ready for discussion, objections, proposals ...

**Comment by ctruta** [comment]
I need to think through the semantics. I'm open to ideas, but on a second thought, I don't know if it's wise to change the behavior of libpng-1.6.
I would personally like to issue an error message for invalid filter values when the filter method is zero. I do not like it when the library silently accepts invalid user input and the user doesn't realize that bad input has silently been changed to some thing valid and arbitrary.
This is something to be changed in the follow-on major versions, however. There is a lot of legacy in libpng-1.6.x, many programs depend on it in unknown ways, and it has been traditional in libpng-land to only apply such changes at major upgrades.
I will prepare, some time next year, for libpng-2.0. That would be a good time for this kind of change.

**Comment by a-kulakov** [comment]
> I don't know if it's wise to change the behavior of libpng-1.6.

But it's a clear bug with PNG_FILTER_VALUE_*.
This bug should/have to be fixed anyway, don't it?

This exactly the case when unexpected results creep silently.
In my case I was forced to switch to libpng16, then discovered noticeable increase in png size and overall performance drop in comparison with libpng15 results. Ok, those, who understand what's going on, can easily fix regress with replacing PNG_FILTER_VALUE_* with single bit mask.

PNG_FILTER_VALUE_* bug fix can be a perfect point to introduce some policy to undocumented values.
Undocumented value results already differ with different libpng versions.
Can be proved with logs from runs of https://github.com/a-kulakov/libpng_set_filter_check code
```BRANCHES='libpng12 linpng15 libpng16' \
EXTRA_CHECKS=ON ./parent_dir_build.sh
```
Can qoute logs here. libpng17 needs extra coding, png_set_filter() became a macro, out of box does not provide any warnings/errors, etc.

**Comment by a-kulakov** [comment]
I've added some extra processing for libpng17 in https://github.com/a-kulakov/libpng_set_filter_check
So far I understand with libpng17 and undocumented value filter png_set_filter() macro silently left previous setting (zero initialization, etc.).
png_setting() returns error.

Will libpng-2.0 follow model with png_setting() ?
I hope libpng-2.0 will not at least use png_set_filter() macro to provide some backward compatibility.

**Comment by ctruta** [comment]
My plan with libpng-2.0 is to follow libpng-1.6 as closely as possible, with minimal disruption to the libpng users. Discard things that I cannot discard in 1.6, and fix things that I cannot fix in 1.6, with the hope that users will be happy to recompile with minimal fuss.

I do not intend to follow the experimental 1.7 branch. It was a good experiment for its time, but it was a step towards API instability, to which I no longer agree. Also, it has too many macros for my taste. I personally consider that branch to be an evolutionary dead end.

The filter check, which you are proposing, should be a good candidate for libpng-2.0.

**Comment by a-kulakov** [comment]
Cosmin, thanks for detailed comment and explanations.
I can only welcome you intention to use libpng16 branch as base for development and share your view on macros. Appreciate a tons of work invested in libpng17, but sometimes was surprised by development decisions and pieces of code I've seen so far :-)

Returning to libpng16 and png_set_filter() bug.
Ok, let's consider too radical fix that set PNG_FILTER_NONE on any undocumented filters values.
But what a set of resulting filters should be ?
Restore v1.6.17 state (same as with libpng15) ?
```
libpng version 1.6.17
 in v1.2 out (hex values, v1.2 - libpng12 output value)
  0  08  08 
  1  10  10 
  2  20  20 
  3  40  40 
  4  80  80 
filter 0x5 libpng warning: Unknown row filter for method 0
  5  00  08 !=
filter 0x6 libpng warning: Unknown row filter for method 0
  6  00  08 !=
filter 0x7 libpng warning: Unknown row filter for method 0
  7  00  08 !=
  8  08  08 
 10  10  10 
  9  09  09 
  a  0a  0a 
100  08  08 
101  10  10 
filter 0x105 libpng warning: Unknown row filter for method 0
105  00  08 !=
filter 0x106 libpng warning: Unknown row filter for method 0
106  00  08 !=
filter 0x107 libpng warning: Unknown row filter for method 0
107  00  08 !=
108  08  08 
110  10  10 
109  09  09 
  0  08  08 
```
Both ```(filters & (PNG_ALL_FILTERS | 0x07))``` and ```(png_byte)``` cast (later sort of coincidentally) just cut higher bits from input. 
Questionable are values 0x9, 0xa and other inputs with mix of mask and value.
Which passed to further processing without changes and without warning/error message.
0x9=(PNG_FILTER_NONE | PNG_FILTER_VALUE_SUB)
0xa=(PNG_FILTER_NONE | PNG_FILTER_VALUE_UP)
...
To limit/eliminate unexpected results from code change in following mask processing,
png_set_filtsers() there should have clear postcondition
```
(png_ptr->do_filter & ~PNG_ALL_FILTERS) == 0
```

**Comment by tangyaofang** [comment]
I agree with a-kulakov that this obvious bug should be removed in the latest version 1.6.XX.
Do you have any plans to modify this bug?
